File: slot-dart/lib/src/views/icon_view.dart
library slot.views.icon_view;

import 'package:stagexl/stagexl.dart';

import 'package:slot/src/mvc/view.dart';
import 'package:slot/src/core/engine.dart';
import 'package:slot/src/models/icon_model.dart';

class IconView extends View {

  Bitmap _image;

  IconView(Model model) : super(model) {

    _createIconImage();

  }

  void _createIconImage() {
    TextureAtlas textureAtlas = Engine.game.assets.getTextureAtlas('atlas01');
    BitmapData bitmapData = textureAtlas.getBitmapData(
        textureAtlas.frames
          .where((f) => f.name.startsWith('icon'))
          .toList()[model.id].name
    );
    bitmapData.renderTexture.filtering = RenderTextureFiltering.NEAREST;
    _image = new Bitmap(bitmapData);
    addChild(_image);

  }

  IconModel get model {
    return super.model;
  }

}

File: slot-ts/src/views/IconView.ts
import Model from '../engine/mvc/Model';
import View from '../engine/mvc/View';
import IconModel from '../models/IconModel';
import SlotAssets from '../SlotAssets';
import Bitmap from 'openfl/display/Bitmap';

export default class IconView extends View {
  private iconBitmap: Bitmap | null = null;

  constructor(model: Model) {
    super(model);
    this.createIconImage();
  }

  protected onUpdated(): void {
    super.onUpdated();
    this.removeChildren();
    this.createIconImage();
  }

  protected createIconImage(): void {
    const iconBitmaps = SlotAssets.getBitmaps('icon_');
    if (this.getModel().getId() < iconBitmaps.length - 1) {
      this.iconBitmap = iconBitmaps[this.getModel().getId()];
      this.addChild(new Bitmap(this.iconBitmap.bitmapData));
    } else {
      throw new Error('Icon texture not found.');
    }
  }

  public getModel(): IconModel {
    return this.model as IconModel;
  }
}
